libpromises/rlist.c

```diff
@@ -527,6 +542,17 @@ Rlist *RlistAppend(Rlist **start, const void *item, RvalType type)
         break;
 
     case RVAL_TYPE_LIST:
+        if (allow_all_types)
+        {
+            JsonElement* store = JsonArrayCreate(RlistLen(item));
+            for (const Rlist *rp = item; rp; rp = rp->next)
+            {
+                JsonArrayAppendElement(store, RvalToJson(rp->val));
+            }
+
+            return RlistAppendRval(start, (Rval) { store, RVAL_TYPE_CONTAINER });
+        }
+
```

Comment:
The old code doesn't collect Rlist elements. So you call `and(splitstring("a b"))` and it ends up as `and("a", "b")`. This expansion happens once per FnCall, not for all arguments.

With the new code, you can optionally (for `COLLECTING` functions) collect all the results of a FnCall into a single JSON array object. It requires functions to expect it, but it's the only reasonable way to preserve backwards compatibility for functions like `and()` in the example above while nesting general calls to functions that take and return compound data types.